File: juyomo/hw3_hw4_1-pointers/48_happy_num.cpp
// Author: Juyoung Moon
// https://leetcode.com/problems/happy-number/

// HOCO Mentoring HW 3
// https://github.com/juyomo/hoco-mentoring

class Solution {
private:
    int getSumOfDigitsSquared(int n) {
        int res = 0;
        while (n > 0) {
            int tmp = n % 10;
            res += (tmp * tmp);
            n = n / 10;
        }
        return res;
    }
public:
    bool isHappy(int n) {
        unordered_set<int> seen;
        while (true) {
            seen.insert(n);
            n = getSumOfDigitsSquared(n);
            if (n == 1) {
                return true;
            }
            if (seen.find(n) != seen.end()) {
                return false;
            }
        }
        return false;
    }
};

File: Jove.L_hoco/week3/48.happy_num.py
'''
한번이라도 등장하면 저장해두고 겹치는지 확인
나온게 또 나오면 무한루프니까 종료
최종 값이 1인지만 확인하자
'''
class Solution:
    def isHappy(self, n: int):
        seen = set()

        while n != 1 and n not in seen:
            seen.add(n)
            
            current_sum = 0
            temp_n = n
            while temp_n > 0:
                digit = temp_n % 10
                current_sum += digit * digit
                temp_n //= 10
            
            n = current_sum
            
        return n == 1